oeps/oep-0022-bp-django-caches.rst

```diff
@@ -0,0 +1,248 @@
++-----------------+--------------------------------------------------------+
+| Type            | Best Practice                                          |
++-----------------+--------------------------------------------------------+
+| Created         | 2018-05-04                                             |
++-----------------+--------------------------------------------------------+
+| `Review Period` | 2018-06-06                                             |
++-----------------+--------------------------------------------------------+
+
+Context
+-------
+
+We have a variety of tools for caching in Django in various repositories,
+but the solutions are not in a shared repository and don't meet all needs. A
+variety of issues were uncovered while working on caching fixes and
+improvements within one of our services.
+
+1. There is a common caching bug while using Django caches where someone
+   caches a value that is Falsey (e.g., None or []) and it is mistakenly
+   treated as a cache miss.
+
+2. In Django we commonly use the Memcached back-end. There are various
+   places in code where we read the same value many times within a
+   the scope of a request, and each of these is a network call to
+   Memcached.
+
+3. For testing and business purposes, it is sometimes useful to be able
+   to invalidate a cached value on a per-request basis.
+
+4. While refactoring caching code and reviewing potential libraries, there
+   seem to be potential best practices that would hide the boilerplate
+   caching logic and better highlight the business logic.
+
+5. There is potential for better standardizing how we handle cache keys.
+   For example, to generate keys that work with Memcached and avoid key
+   collisions in a global space like Memcached.
+
+Note: When and if caching is warranted in any particular circumstance is beyond
+the scope of this OEP.
+
+Recommendations
+---------------
+
+The following recommendations include patterns that are generally useful. The
+intention is to better codify these recommendations into shared libraries over
+time and update this documentation as the need arises.
+
+Common Caching Defect and Fix
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The following is a common defect when caching using Django caches::
+
+    cached_value = cache.get(key)
+    if not cached_value:
+        # calculate value, set in cache, and return
+        # if the calculated value is Falsey (e.g., None or []), it will
+        # be treated as a cache miss above.
+    return cached_value
+
+The simplest fix for this defect is::
+
+    _CACHE_MISS = object()
+
+    cached_value = cache.get(key, _CACHE_MISS)
+    if cached_value is _CACHE_MISS:
+        # calculate value, set in cache, and return
+        # if the calculated value is Falsey (e.g., None or []), it will
+        # no longer be treated as a cache miss above.
+    return cached_value
+
+Use libraries that makes it simpler to avoid the above defect. See
```

Comment:
I don't understand this sentence (and it is slightly non-grammatical).

Reply:
Is this more clear?:
```
Where possible, use caching utility methods that avoid this defect.
```

Reply:
I think what I was getting at was: there's a switch here from "Problem statement" to "Solution" that is sudden and implicit.  Maybe, "To solve this problem, use libraries that ...."